### ISSUE TITLE
formatter breaks Liquid in Javascript

### ISSUE THREAD

**Issue by resistorsoftware** [panoply/vscode-liquid#11]
Version 1.6.5 of the plugin.

I notice at times if I write javascript in a .liquid file (which the bottom status bar continues to tell me is HTML) the formatting will have problems on some but not all Liquid. 

**Expected Style**

var fizzbuzz = function() {
   var foobazz = {{ some_variable }};
}

** Resulting Format **

var fizzbuzz = function() {
   var foobazz = {{ some_variable }
};
}

So in a larger, non-trivial template, this kind of goof can be frustrating to track down, as it mangles not only the liquid, but the javascript as well. 

I am looking for some insight as to how to setup VSCode to allow me to work on Liquid templates with Javascript in them, and not have this particular formatting gremlin be a thing.

**Comment by panoply** [comment]
@resistorsoftware the extension doesn't support liquid infused JavaScript. Are you writing this JavaScript within a `<script>` tag? The formatter will show when the file uses a `.liquid` and `.html` extension and it will attempt to format these files as if they where HTML.

Would have the ability to exclude certain files from formatting be helpful in this situation? – I am working on a enabling developers to include a `.liquirc` file in their project which would contain custom options and the extension will use and apply opposed to the current method of defining options in user settings. This feature would allow you to exclude (ignore) certain files from being formatted.

Just out of curiosity, what is your use case here? There are always alternative ways to reference liquid specific references opposed to infusing them within your js.

**Comment by resistorsoftware** [comment]
I write javascript that manipulates the DOM of Shopify templates. The typical use case is when some special value is encoding in Liquid and I need that value in the javascript to ensure the UX performs as needed. So in my case, I am seeing this formatting glitch in the javascript, which you point out, might be 100% due to the HTML formatter and not anything in the liquid extension.

Say I have a server-side value I need, it will encoded as:

    product.metafields.namespace.key 

So I save that as a liquid variable using an **assign**. Now in my javascript I need that assignment and so I might use:

    var myVar = {{ liquidVar }};

which as I say, gets mangled sometimes by a **formatter** to:

    var myVar = {{ liquidVar }
    };

Thanks for your quick input on this! Much appreciated! I will keep careful notes of what I see so that I get hopefully drill town to the responsible party!

**Comment by panoply** [comment]
I know what you're trying to achieve, in cases like this when working with Shopify more specifically it's better to define the server side requirements in global accessible object and just reference when required in your bundle, for example:

```html
<script>
var theme = {
	product: {
		namespace: {{ product.metafields.namespace.key }},
		ref2: {{ ref2 }}
	}
}
</script>
```

As long as you call this before your compiled bundle, the `theme.product.namespace` is accessible in your bundle. You can also use data attributes and just access the data right from the node.

**Comment by resistorsoftware** [comment]
I am not following the logic here. If the formatter is breaking js.liquid, why does it matter where that is defined? I am not familiar with the lingo here, specifically what you mean by **bundle**? 

I localize my javascript to sections of code where it is needed. It would be hard, perhaps even impossible to extract the javascript elsewhere, populate it with server-side data, and then access the instantiated javascript in my custom templates without some pain points. 

In other words, with what you are telling me, I am not seeing how that <script> you wrote is immune to being mangled? What makes that <script> pass through the formatter without the possibility of it turning out wrong like:

    <script>
      var theme = {
	product: {
		namespace: {{ product.metafields.namespace.key }
         },
		ref2: {{ ref2 }}
	 }
       }
    </script>

**Comment by panoply** [comment]
It's just a recommendation for your theme development, an alternative approach opposed to having liquid infused JavaScript that is transpiled by Shopify. It's unrelated to the extension.

I can exclude `.js.liquid` files from being formatted from now to prevent the extension from formatting these extension types, keep in mind syntax highlighting will also be disabled when this is applied.

**Comment by panoply** [comment]
You can just use file associations for `.js.liquid` files in your `settings.json`, eg: 

```json
  "files.associations": {
    "*.js.liquid": "javascript"
  },
```

**Comment by resistorsoftware** [comment]
Interesting. I recently encountered a head scratcher that I still don't quite grok, but it seems related. I had a error in a *.js.liquid file where the following line was bombing:

    theme.Currency = (function() {
      var moneyFormat = "${{amount}}";

The error was that moneyFormat was "$". Strange. It should have been a string "${{amount}}" but Shopify compiled the Liquid and since amount is not a thing, it resulted in the "$". WTF is that! Wrong! And this is stock Shopify code. I am lost as to what they meant. I get it that the .js.liquid allows one to mix JS and Liquid and thus this works.. but the Liquid render part should not have kicked in.

The only fix I came up with was the atom bomb... using {% raw %}... this part of the Shopify tech stack is often mysterious and undocumented. 

    theme.Currency = (function() {
      var moneyFormat = "${%raw%}{{amount}}{%endraw%}";

This correctly populated moneyFormat with "${{amount}}"  but I have no clue why I even had to do this.

**Comment by panoply** [comment]
If memory calls me correct `${{amount}}` needs to be called inline using a script tag. Because Shopify transpiles `.js.liquid` and `scss.liquid` files at runtime it will treat liquid variables as well variables. The `*.js.liquid` files should really only contain references.

I could be wrong here but I'm quite sure the `${{amount}}` tag is actually handled by some of the scripts which are injected via the `{{ content_for_header }}` tag and if you're calling it from within a `.js.liquid` files this specific tag cannot be handled. 

Did file association resolve the issue for you?

**Comment by resistorsoftware** [comment]
I added the association and now when I edit the js.liquid files the association is indeed Javascript. Now I will watch out for the so-called **mangling** effect! By that I mean I want to see if happens on copy&paste operations, etc.

**Comment by panoply** [comment]
If the Liquid toggle button is not shown then the formatter will not be active. Closing.

**Comment by resistorsoftware** [comment]
I never see a Liquid toggle button for this extension no matter what file I edit. The plugin is installed and yet this is never an option.

**Comment by panoply** [comment]
Your issues all seem unrelated to this extension. Verify you're using the latest version and ensure no other extensions are causing conflict. If you don't see the toggle button in the status bar but you are still able to format files, the extension is working. File associations will disable to formatter on `*.js.liquid` extension files. Go over the [VSCode Docs](https://code.visualstudio.com/Docs) for a refresher so you can easily diagnose other issues that might arise.
